File: bridge-tmux/internal/tmux/session.go
// Package tmux manages interactive Claude Code sessions via tmux (Tier2).
//
// 各 peer に 1 つの tmux セッションを割り当て、以下を行う:
//   - Start(): tmux new-session + claude 起動コマンド送信
//   - Stop(): Ctrl+C → 5 秒後に kill-session
//   - InjectMessage(): named buffer 経由でメッセージを貼り付け + Enter
//   - WaitForIdle(): pane 変化ゼロ N 秒 → 応答完了
//
// 応答は claude が MCP tool (send_message) 経由で自律的に送信する。
// bridge (Tier1) は応答テキストを解析する必要がない。
//
// Issue: #110
package tmux

import (
	"context"
	"fmt"
	"io"
	"log"
	"os/exec"
	"sort"
	"strings"
	"time"
)

const (
	defaultPollInterval    = 500 * time.Millisecond
	defaultGracefulWait    = 5 * time.Second
	defaultMinActivityWait = 1 * time.Second
)

// tmuxRunner abstracts tmux exec.Command calls for testability.
// Production uses realRunner; tests inject a mock.
type tmuxRunner interface {
	run(args ...string) error
	runCtx(ctx context.Context, args ...string) error
	runWithStdin(stdin io.Reader, args ...string) error
	output(args ...string) ([]byte, error)
}

// SessionIface is the interface implemented by *Session.
// Use it to inject mock sessions in tests (e.g. in SessionManager).
type SessionIface interface {
	IsAlive() bool
	Start(ctx context.Context) error
	Stop(ctx context.Context) error
	InjectMessage(text string) error
	WaitForIdle(ctx context.Context) error
}

// Compile-time assertion: *Session must implement SessionIface.
var _ SessionIface = (*Session)(nil)

// realRunner is the production tmuxRunner using exec.Command.
type realRunner struct{}

func (realRunner) run(args ...string) error {
	return exec.Command("tmux", args...).Run()
}

func (realRunner) runCtx(ctx context.Context, args ...string) error {
	return exec.CommandContext(ctx, "tmux", args...).Run()
}

func (realRunner) runWithStdin(stdin io.Reader, args ...string) error {
	cmd := exec.Command("tmux", args...)
	cmd.Stdin = stdin
	return cmd.Run()
}

func (realRunner) output(args ...string) ([]byte, error) {
	return exec.Command("tmux", args...).Output()
}

// Session は 1 peer に対応する tmux セッション。
type Session struct {
	Name             string
	Workdir          string
	MCPConfigPath    string
	ClaudeCLI        string
	Model            string
	BypassPerms      bool
	startedBefore    bool
	SpawnTimeout     time.Duration
	ActivityIdleTime time.Duration
	ResponseTimeout  time.Duration
	// Env は Tier2 (claude) 起動時に追加設定する環境変数。
	// fleet mode で persona ごとに異なる env を渡す用途。
	// キーはソートされてコマンド先頭に KEY='val' 形式で付加される。
	Env map[string]string
	// Timing overrides — zero means use default constant.
	PollInterval    time.Duration // default: 500ms
	GracefulWait    time.Duration // default: 5s
	MinActivityWait time.Duration // default: 1s
	runner          tmuxRunner
}

// SessionOptions は NewSession のオプション。
type SessionOptions struct {
	UserID           string
	Workdir          string
	MCPConfigPath    string
	ClaudeCLI        string
	Model            string
	BypassPerms      bool
	SpawnTimeout     time.Duration
	ActivityIdleTime time.Duration
	ResponseTimeout  time.Duration
	// Env は persona ごとの追加環境変数 (fleet mode 用)。
	Env map[string]string
}

// NewSession は Session を生成する (tmux セッションはまだ作らない)。
func NewSession(opts SessionOptions) *Session {
	name := "claude-bridge-" + opts.UserID
	return &Session{
		Name:             name,
		Workdir:          opts.Workdir,
		MCPConfigPath:    opts.MCPConfigPath,
		ClaudeCLI:        opts.ClaudeCLI,
		Model:            opts.Model,
		BypassPerms:      opts.BypassPerms,
		SpawnTimeout:     opts.SpawnTimeout,
		ActivityIdleTime: opts.ActivityIdleTime,
		ResponseTimeout:  opts.ResponseTimeout,
		Env:              opts.Env,
		runner:           realRunner{},
	}
}

func (s *Session) pollInterval() time.Duration {
	if s.PollInterval > 0 {
		return s.PollInterval
	}
	return defaultPollInterval
}

func (s *Session) gracefulWait() time.Duration {
	if s.GracefulWait > 0 {
		return s.GracefulWait
	}
	return defaultGracefulWait
}

func (s *Session) minActivityWait() time.Duration {
	if s.MinActivityWait > 0 {
		return s.MinActivityWait
	}
	return defaultMinActivityWait
}

// ──────────────────────────────────────────────────────────────────────── //
// ライフサイクル                                                           //
// ──────────────────────────────────────────────────────────────────────── //

// IsAlive は tmux セッションが存在するか確認する。
func (s *Session) IsAlive() bool {
	return s.runner.run("has-session", "-t", s.Name) == nil
}

// Start は tmux セッションを新規作成して claude を起動する。
// s.startedBefore が true なら --continue で会話を継続する。
func (s *Session) Start(ctx context.Context) error {
	if s.IsAlive() {
		log.Printf("[tmux] session %s already exists — stopping first", s.Name)
		if err := s.Stop(ctx); err != nil {
			return err
		}
	}

	log.Printf("[tmux] creating session %s (workdir=%s)", s.Name, s.Workdir)
	if err := s.runner.runCtx(ctx,
		"new-session", "-d",
		"-s", s.Name,
		"-c", s.Workdir,
	); err != nil {
		return fmt.Errorf("tmux new-session: %w", err)
	}

	cmdStr := s.buildCLICommand()
	log.Printf("[tmux] starting claude: %s", cmdStr)
	if err := s.runner.runCtx(ctx, "send-keys", "-t", s.Name, cmdStr, "Enter"); err != nil {
		// send-keys 失敗時は空セッションが残留しないよう best-effort cleanup する
		_ = s.Stop(ctx)
		return fmt.Errorf("tmux send-keys (start): %w", err)
	}

	// claude が起動して pane に何か出力するまで待つ
	if wait := s.minActivityWait(); wait > 0 {
		time.Sleep(wait)
	}
	deadline := time.Now().Add(s.SpawnTimeout)
	baseline := s.capturePaneText()

	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		content := s.capturePaneText()
		if content != baseline && strings.TrimSpace(content) != "" {
			log.Printf("[tmux] session %s started — got initial output", s.Name)
			s.startedBefore = true
			return nil
		}
		time.Sleep(s.pollInterval())
	}

	_ = s.Stop(ctx)
	return fmt.Errorf("claude did not start within %.0fs in session %s",
		s.SpawnTimeout.Seconds(), s.Name)
}

// Stop は tmux セッションを停止する (graceful → force kill)。
func (s *Session) Stop(ctx context.Context) error {
	if !s.IsAlive() {
		return nil
	}
	log.Printf("[tmux] stopping session %s", s.Name)

	// Ctrl+C で graceful 終了を試みる
	_ = s.runner.run("send-keys", "-t", s.Name, "C-c", "")

	// gracefulWait 待って、まだ生きていれば force kill
	gracefulTimer := time.NewTimer(s.gracefulWait())
	defer gracefulTimer.Stop()
	select {
	case <-gracefulTimer.C:
	case <-ctx.Done():
	}

	if s.IsAlive() {
		_ = s.runner.run("kill-session", "-t", s.Name)
	}
	log.Printf("[tmux] session %s stopped", s.Name)
	return nil
}

// ──────────────────────────────────────────────────────────────────────── //
// メッセージング                                                           //
// ──────────────────────────────────────────────────────────────────────── //

// InjectMessage はプロンプトテキストを tmux ペインに貼り付け Enter を送る。
// named buffer を使うことで、同一ホストで複数 bridge が動いても buffer 競合しない。
func (s *Session) InjectMessage(text string) error {
	bufName := "bridge-" + s.Name

	if err := s.runner.runWithStdin(strings.NewReader(text),
		"load-buffer", "-b", bufName, "-",
	); err != nil {
		return fmt.Errorf("tmux load-buffer: %w", err)
	}

	if err := s.runner.run("paste-buffer", "-b", bufName, "-t", s.Name); err != nil {
		return fmt.Errorf("tmux paste-buffer: %w", err)
	}

	if err := s.runner.run("send-keys", "-t", s.Name, "", "Enter"); err != nil {
		return fmt.Errorf("tmux send-keys (enter): %w", err)
	}

	// buffer を削除 (PAT 等の機密情報をメモリから消す)
	_ = s.runner.run("delete-buffer", "-b", bufName)

	log.Printf("[tmux] injected %d chars to session %s", len(text), s.Name)
	return nil
}

// WaitForIdle は claude の応答完了を待つ。
//
// アルゴリズム:
//  1. pane 変化待ち (claude が処理を開始した証拠)
//  2. pane 変化が止まって ActivityIdleTime 秒経過 → 完了と判断
//  3. ResponseTimeout 超過 → error
func (s *Session) WaitForIdle(ctx context.Context) error {
	deadline := time.Now().Add(s.ResponseTimeout)
	baseline := s.capturePaneText()

	// Phase 1: pane が変化するまで待つ
	activityStarted := false
	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		content := s.capturePaneText()
		if content != baseline {
			activityStarted = true
			baseline = content
			break
		}
		time.Sleep(s.pollInterval())
	}
	if !activityStarted {
		return fmt.Errorf("session %s: claude did not start processing within %.0fs",
			s.Name, s.ResponseTimeout.Seconds())
	}

	// Phase 2: pane 変化が止まるまで待つ
	lastContent := baseline
	lastChange := time.Now()

	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		time.Sleep(s.pollInterval())

		// tmux セッションが死んだ場合は capturePaneText が "" を返し続ける。
		// 「変化なし」→ idle timeout で false success にならないよう死活確認する。
		if !s.IsAlive() {
			return fmt.Errorf("session %s: tmux session died while waiting for response", s.Name)
		}

		content := s.capturePaneText()
		if content != lastContent {
			lastContent = content
			lastChange = time.Now()
		} else if time.Since(lastChange) >= s.ActivityIdleTime {
			log.Printf("[tmux] session %s idle for %.1fs — response complete",
				s.Name, s.ActivityIdleTime.Seconds())
			return nil
		}
	}

	return fmt.Errorf("session %s: response timeout (%.0fs)", s.Name, s.ResponseTimeout.Seconds())
}

// ──────────────────────────────────────────────────────────────────────── //
// 内部実装                                                                 //
// ──────────────────────────────────────────────────────────────────────── //

func (s *Session) capturePaneText() string {
	out, err := s.runner.output("capture-pane", "-p", "-S", "-", "-t", s.Name)
	if err != nil {
		return ""
	}
	return string(out)
}

// buildCLICommand は claude 起動コマンド文字列を組み立てる。
// シェル展開を避けるため各引数を単引用符でエスケープする。
// Env が設定されている場合は "KEY='val'" 形式でコマンド先頭に付加する。
// キーはソートして決定的な順序を保証する。
func (s *Session) buildCLICommand() string {
	var parts []string

	// 追加環境変数をコマンド先頭に付加 (KEY='val' 形式)
	if len(s.Env) > 0 {
		keys := make([]string, 0, len(s.Env))
		for k := range s.Env {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			parts = append(parts, k+"="+shellQuote(s.Env[k]))
		}
	}

	parts = append(parts, shellQuote(s.ClaudeCLI))
	parts = append(parts, "--mcp-config", shellQuote(s.MCPConfigPath))
	if s.startedBefore {
		parts = append(parts, "--continue")
	}
	if s.BypassPerms {
		// --dangerously-skip-permissions は interactive dialog を表示して pane が静止し
		// Start() の spawn タイムアウトを引き起こす。--permission-mode bypassPermissions
		// は dialog なしで即起動する (bridge-claude と同じ方式)。
		parts = append(parts, "--permission-mode", "bypassPermissions")
	}
	if s.Model != "" {
		parts = append(parts, "--model", shellQuote(s.Model))
	}
	return strings.Join(parts, " ")
}

// shellQuote は安全な shell 引用符を付ける (単引用符ベース)。
// 値に単引用符が含まれる場合は置換する (= POSIX 準拠)。
func shellQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "'\\''") + "'"
}

File: src/agent_hub_bridges/tmux/session.py
"""tmux セッション管理 (Tier2) — TmuxSession + SessionManager.

TmuxSession: 1 peer に対する 1 つの tmux セッションのライフサイクルを管理する。
  - 起動 (start): `tmux new-session` + `claude --mcp-config ...` を送信
  - メッセージ注入 (inject_message): `tmux load-buffer` + `paste-buffer` + Enter
  - 応答完了検知 (wait_for_idle): pane 変化ゼロ N 秒で完了と判断
  - 停止 (stop): graceful kill → 5 秒後に force kill

SessionManager: idle タイムアウトによる自動 kill と on-demand 再起動を担う。
  - handle(prompt): 未起動なら start()、inject、wait_for_idle の一連をラップ
  - idle timer: 最終メッセージ処理完了から idle_timeout_s 後に stop()

設計: docs/design-bridge-tmux.md §Tier2
Issue: #110
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import shlex
import shutil
import subprocess
import tempfile
import time
from dataclasses import dataclass, field
from pathlib import Path

from agent_hub_bridges.tmux.config import Config

logger = logging.getLogger(__name__)

# pane 変化ポーリング間隔
_POLL_INTERVAL_S = 0.5

# graceful shutdown 後に force kill するまで待つ秒数
_GRACEFUL_KILL_WAIT_S = 5.0

# cold start 後に pane に何か出力が来るまで待つ最小時間
_MIN_ACTIVITY_WAIT_S = 1.0


@dataclass
class TmuxSession:
    """1 peer に対応する interactive Claude Code tmux セッション.

    Tier2 (claude) の実体。起動・停止・メッセージ注入・応答完了検知を提供する。

    Attributes:
        session_name: tmux セッション名 (例: `claude-bridge-reviewer`)。
        workdir: peer の作業 root (CLAUDE.md が置かれた dir)。
        _config: bridge-tmux 全体の runtime config。
        _mcp_config_path: claude に渡す MCP config 一時ファイル (close() で削除)。
        _cli_path: claude CLI の resolved path。
        _started_before: True なら次の start() で `--continue` を使う。
    """

    session_name: str
    workdir: Path
    _config: Config = field(repr=False)
    _mcp_config_path: Path = field(repr=False)
    _cli_path: str = field(repr=False)
    _started_before: bool = field(default=False, repr=False)

    # ------------------------------------------------------------------ #
    # factory                                                              #
    # ------------------------------------------------------------------ #

    @classmethod
    def create(cls, config: Config) -> TmuxSession:
        """Session オブジェクトを生成する (まだ tmux session は作らない)."""
        cli_path = shutil.which(config.claude_cli_path) or config.claude_cli_path
        if not Path(cli_path).exists() and not shutil.which(cli_path):
            raise FileNotFoundError(
                f"claude CLI not found: '{config.claude_cli_path}'. "
                "Install Claude Code and make sure it's on PATH."
            )
        mcp_config_path = _write_mcp_config(config)
        session_name = f"claude-bridge-{config.user}"
        logger.info(
            "TmuxSession created: name=%s workdir=%s cli=%s",
            session_name, config.workdir, cli_path,
        )
        return cls(
            session_name=session_name,
            workdir=config.workdir,
            _config=config,
            _mcp_config_path=mcp_config_path,
            _cli_path=cli_path,
        )

    # ------------------------------------------------------------------ #
    # lifecycle                                                            #
    # ------------------------------------------------------------------ #

    def is_alive(self) -> bool:
        """tmux セッションが存在するか確認する."""
        result = subprocess.run(
            ["tmux", "has-session", "-t", self.session_name],
            capture_output=True,
        )
        return result.returncode == 0

    async def start(self) -> None:
        """tmux セッションを新規作成して claude を起動する.

        _started_before が True なら `--continue` で直前の会話を継続する。
        起動後 spawn_timeout_s 以内に pane に出力が現れるのを待つ。
        """
        if self.is_alive():
            logger.warning("Session %s already exists — stopping first", self.session_name)
            await self.stop()

        # tmux セッション作成 (detached)
        logger.info("Creating tmux session %s (workdir=%s)", self.session_name, self.workdir)
        subprocess.run(
            [
                "tmux", "new-session", "-d",
                "-s", self.session_name,
                "-c", str(self.workdir),
            ],
            check=True,
        )

        # claude 起動コマンドを送信
        cmd_parts = self._build_cmd()
        cmd_str = " ".join(shlex.quote(p) for p in cmd_parts)
        subprocess.run(
            ["tmux", "send-keys", "-t", self.session_name, cmd_str, "Enter"],
            check=True,
        )
        logger.info("Sent claude start command to %s: %s", self.session_name, cmd_str)

        # pane に出力が来るまで待機 (= claude が起動した証拠)
        deadline = time.monotonic() + self._config.spawn_timeout_s
        baseline = self._capture_pane()

        # send-keys した直後はコマンド文字列が pane に出るので少し待ってから判定
        await asyncio.sleep(_MIN_ACTIVITY_WAIT_S)

        while time.monotonic() < deadline:
            content = self._capture_pane()
            if content != baseline and content.strip():
                logger.info("Session %s started — got initial output", self.session_name)
                self._started_before = True
                return
            await asyncio.sleep(_POLL_INTERVAL_S)

        # タイムアウト: セッションは起動したが claude から出力なし
        logger.error(
            "Session %s spawn timeout (%.0fs)", self.session_name, self._config.spawn_timeout_s
        )
        await self.stop()
        raise TimeoutError(
            f"claude did not produce output within {self._config.spawn_timeout_s:.0f}s "
            f"in tmux session '{self.session_name}'"
        )

    async def stop(self) -> None:
        """tmux セッションを停止する (graceful → force)."""
        if not self.is_alive():
            return
        logger.info("Stopping tmux session %s", self.session_name)

        # Ctrl+C を送って graceful 終了を試みる
        subprocess.run(
            ["tmux", "send-keys", "-t", self.session_name, "C-c", ""],
            check=False,
        )
        await asyncio.sleep(_GRACEFUL_KILL_WAIT_S)

        # まだ生きていれば force kill
        if self.is_alive():
            subprocess.run(
                ["tmux", "kill-session", "-t", self.session_name],
                check=False,
            )
        logger.info("tmux session %s stopped", self.session_name)

    def close(self) -> None:
        """リソースのクリーンアップ (MCP config 一時ファイル削除)."""
        try:
            self._mcp_config_path.unlink(missing_ok=True)
        except Exception:
            logger.warning("Failed to delete MCP config: %s", self._mcp_config_path)

    # ------------------------------------------------------------------ #
    # messaging                                                            #
    # ------------------------------------------------------------------ #

    async def inject_message(self, text: str) -> None:
        """メッセージテキストを tmux ペインに paste して Enter を送る.

        Named buffer (`bridge-<session_name>`) を使うことで、
        同一ホストで複数 bridge が動いても global buffer が競合しない。
        """
        buffer_name = f"bridge-{self.session_name}"
        try:
            # テキストをバッファに書き込む
            subprocess.run(
                ["tmux", "load-buffer", "-b", buffer_name, "-"],
                input=text.encode("utf-8"),
                check=True,
            )
            # バッファをペインに貼り付ける
            subprocess.run(
                ["tmux", "paste-buffer", "-b", buffer_name, "-t", self.session_name],
                check=True,
            )
            # Enter を送信
            subprocess.run(
                ["tmux", "send-keys", "-t", self.session_name, "", "Enter"],
                check=True,
            )
        finally:
            # 使い終わったバッファを削除 (PAT 等の機密情報をメモリから消す)
            subprocess.run(
                ["tmux", "delete-buffer", "-b", buffer_name],
                check=False,
            )
        logger.info(
            "Injected message to session %s (%d chars)",
            self.session_name, len(text),
        )

    async def wait_for_idle(self) -> None:
        """claude の応答完了を待つ.

        アルゴリズム:
          1. 注入直後の pane 内容を baseline として記録する
          2. pane が変化するまでポーリング (= claude が応答を始めた)
          3. pane 変化が止まって activity_idle_s 秒経過 → 完了と判断
          4. response_timeout_s 超過 → TimeoutError

        応答は claude が MCP tool (send_message) 経由で送信するため、
        bridge 側は pane テキストを解析する必要はない。
        idle 検知後に worker が hub.ack() を呼ぶだけでよい。
        """
        activity_idle_s = self._config.activity_idle_s
        response_timeout_s = self._config.response_timeout_s
        deadline = time.monotonic() + response_timeout_s

        baseline = self._capture_pane()
        activity_started = False
        last_change_time = time.monotonic()
        last_content = baseline

        # Phase 1: pane 変化待ち (claude が処理を開始した証拠)
        while time.monotonic() < deadline:
            await asyncio.sleep(_POLL_INTERVAL_S)
            content = self._capture_pane()
            if content != baseline:
                activity_started = True
                last_content = content
                last_change_time = time.monotonic()
                logger.debug("Session %s: activity started", self.session_name)
                break

        if not activity_started:
            raise TimeoutError(
                f"Session {self.session_name}: claude did not start processing "
                f"within {response_timeout_s:.0f}s"
            )

        # Phase 2: pane 変化が止まるまで待つ
        while time.monotonic() < deadline:
            await asyncio.sleep(_POLL_INTERVAL_S)
            content = self._capture_pane()
            if content != last_content:
                last_content = content
                last_change_time = time.monotonic()
            elif time.monotonic() - last_change_time >= activity_idle_s:
                logger.info(
                    "Session %s: idle for %.1fs — response complete",
                    self.session_name, activity_idle_s,
                )
                return

        raise TimeoutError(
            f"Session {self.session_name}: response timeout ({response_timeout_s:.0f}s)"
        )

    # ------------------------------------------------------------------ #
    # internals                                                            #
    # ------------------------------------------------------------------ #

    def _capture_pane(self) -> str:
        """tmux capture-pane -p で pane のテキストを取得する."""
        result = subprocess.run(
            ["tmux", "capture-pane", "-p", "-S", "-", "-t", self.session_name],
            capture_output=True,
            text=True,
        )
        return result.stdout

    def _build_cmd(self) -> list[str]:
        """claude 起動コマンドを組み立てる.

        _started_before が True なら --continue を付けて直前の会話を継続する。
        ANTHROPIC_API_KEY はスタートアップスクリプト経由で unset する。
        """
        cmd = [self._cli_path, "--mcp-config", str(self._mcp_config_path)]
        if self._started_before:
            cmd.append("--continue")
        if self._config.permission_bypass:
            cmd.append("--dangerously-skip-permissions")
        if self._config.model:
            cmd.extend(["--model", self._config.model])
        return cmd


# ──────────────────────────────────────────────────────────────────────── #
# SessionManager: on-demand spawn + idle timer                            #
# ──────────────────────────────────────────────────────────────────────── #


class SessionManager:
    """on-demand spawn と idle タイムアウト kill を担うセッションマネージャ.

    TmuxSession のライフサイクル (Cold -> Warm -> Cooling -> Cold) を管理し、
    worker からは `handle(prompt)` を呼ぶだけでよい抽象を提供する。

    Attributes:
        _session: 管理対象の TmuxSession。
        _idle_timer: idle タイムアウト後に stop() を呼ぶ asyncio.Task。
        _lock: 同一セッションへの同時アクセスを防ぐ Mutex。
    """

    def __init__(self, session: TmuxSession) -> None:
        self._session = session
        self._idle_timer: asyncio.Task[None] | None = None
        self._lock = asyncio.Lock()

    @classmethod
    def create(cls, config: Config) -> SessionManager:
        """SessionManager を生成する."""
        session = TmuxSession.create(config)
        return cls(session)

    async def handle(self, prompt: str) -> None:
        """メッセージを処理する.

        Cold なら Tier2 を spawn してから inject → wait_for_idle。
        完了後は idle timer をセット (idle_timeout_s 後に kill)。
        """
        async with self._lock:
            # idle timer をキャンセル (新メッセージ到着で wake)
            self._cancel_idle_timer()

            # Cold なら Tier2 を起動する
            if not self._session.is_alive():
                logger.info(
                    "Session %s is cold — spawning Tier2", self._session.session_name
                )
                await self._session.start()

            # メッセージを注入する
            await self._session.inject_message(prompt)

        # wait_for_idle は lock 外で実行 (長時間 block しないよう)
        try:
            await self._session.wait_for_idle()
        except TimeoutError as exc:
            logger.error("Response timeout for session %s: %s", self._session.session_name, exc)
            # タイムアウト時はセッションをリセットして次のメッセージに備える
            async with self._lock:
                await self._session.stop()
            raise

        async with self._lock:
            # idle timer を開始する
            self._start_idle_timer()

    def close(self) -> None:
        """リソースのクリーンアップ (idle timer キャンセル + MCP config 削除)."""
        self._cancel_idle_timer()
        self._session.close()

    async def shutdown(self) -> None:
        """セッションを graceful に停止する (bridge 終了時)."""
        self._cancel_idle_timer()
        await self._session.stop()
        self._session.close()

    # ------------------------------------------------------------------ #
    # idle timer                                                           #
    # ------------------------------------------------------------------ #

    def _start_idle_timer(self) -> None:
        self._idle_timer = asyncio.create_task(
            self._run_idle_timer(),
            name=f"idle-timer-{self._session.session_name}",
        )

    def _cancel_idle_timer(self) -> None:
        if self._idle_timer and not self._idle_timer.done():
            self._idle_timer.cancel()
            self._idle_timer = None

    async def _run_idle_timer(self) -> None:
        """idle_timeout_s 後に Tier2 を kill する."""
        try:
            await asyncio.sleep(self._session._config.idle_timeout_s)
            logger.info(
                "Idle timeout (%.0fs) — killing session %s",
                self._session._config.idle_timeout_s,
                self._session.session_name,
            )
            await self._session.stop()
        except asyncio.CancelledError:
            pass  # 新メッセージ到着でキャンセルされた (正常)


# ──────────────────────────────────────────────────────────────────────── #
# MCP config 生成                                                          #
# ──────────────────────────────────────────────────────────────────────── #


def _write_mcp_config(config: Config) -> Path:
    """claude --mcp-config に渡す一時 JSON ファイルを書く.

    claude_p bridge と同一形式。mode 0o600 で GITHUB_PAT を保護する。
    `TmuxSession.close()` 時に削除される。

    Issue: #84 caused_by 因果チェーン対応のため X-User-Id ヘッダを送る。
    """
    headers: dict[str, str] = {
        "Authorization": f"Bearer {config.github_pat}",
        "X-User-Id": config.user,
    }
    if config.tenant:
        headers["X-Tenant-Id"] = config.tenant

    mcp_payload = {
        "mcpServers": {
            "agent-hub": {
                "type": "http",
                "url": config.agent_hub_url,
                "headers": headers,
            }
        }
    }

    fd, tmp_path_str = tempfile.mkstemp(
        suffix=".json", prefix=f"bridge-tmux-{config.user}-"
    )
    tmp_path = Path(tmp_path_str)
    try:
        os.chmod(tmp_path, 0o600)
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            json.dump(mcp_payload, f, indent=2)
    except Exception:
        tmp_path.unlink(missing_ok=True)
        raise

    logger.info("Wrote MCP config: %s", tmp_path)
    return tmp_path
